src/Functions/LowerUpperUTF8Impl.h

```diff
@@ -19,71 +18,7 @@ namespace ErrorCodes
     extern const int BAD_ARGUMENTS;
 }
 
-/// xor or do nothing
-template <bool>
-UInt8 xor_or_identity(const UInt8 c, const int mask)
-{
-    return c ^ mask;
-}
-
-template <>
-inline UInt8 xor_or_identity<false>(const UInt8 c, const int)
-{
-    return c;
-}
-
-/// It is caller's responsibility to ensure the presence of a valid cyrillic sequence in array
-template <bool to_lower>
-inline void UTF8CyrillicToCase(const UInt8 *& src, UInt8 *& dst)
```

Comment:
A fast-path for Latin+Cyrillic should remain. Add a performance test.

Reply:
It is tested manully in https://github.com/ClickHouse/ClickHouse/pull/65761#discussion_r1660435973 and https://github.com/ClickHouse/ClickHouse/pull/65761#discussion_r1660438004     The new implementation has little impact on perforamance.